### ISSUE TITLE
Re-enable some of the epydoc features

### ISSUE THREAD

**Issue by tristanlatr** [twisted/pydoctor#298]
Hello,

I saw that last year, there was a big change in epydoc code, everything has been imported and removed part by part. Which is really cool to review the changes.

There is a few cool features that were removed. Mostly because it was not used by the twisted community. 

I think the features regarding graph generation should be supported. I'm generating graph with pyreverse, but it would be awsome to have the package and the class UML inside pydoctor on demand. 947c39d7c0453d3ed831269da2b196572127bed1 and 9e53d5b34cd26d0a41a10f269611b2d89fb2446f

Also the documentation completeness checker aa309cf852946ab6b6eea2d647d5d9f6e9e2cdbe It's a cool idea, and I'm sure there are people who will use it

The latex generator 9e53d5b34cd26d0a41a10f269611b2d89fb2446f , this is to output the whole thing as PDF?

**Comment by mthuurne** [comment]
> There is a few cool features that were removed. Mostly because it was not used by the twisted community.

Back then, I looked at Twisted to determine which features are essential and removed everything non-essential, as the Python 2.7 deadline was looming and our dependency on epydoc was blocking the port to Python 3. So I agree that if some of those features are useful, they can be brought back now.

While in one of the commit comments I suggested reverting it to bring back the feature, I'm now not so sure that would actually work. When it comes to parsing docstrings, our code still resembles epydoc a lot. But when it comes to analyzing source code, it has always been different and that difference has only grown over time as pydoctor evolved with the Python language and common practices (type annotations, `attrs` etc.) while epydoc stood still.

For example, for checking completeness, I think it would be easier to introduce an option that warns whenever an "Undocumented" phrase is inserted into the output instead of re-introducing epydoc's checker module. Otherwise we'd have problems for example deciding when to warn about a missing `type` field: if there is a type annotation, there is no need warn, but epydoc's code doesn't have access to the information we extract from annotations.

> I think the features regarding graph generation should be supported. I'm generating graph with pyreverse, but it would be awsome to have the package and the class UML inside pydoctor on demand. [947c39d](https://github.com/twisted/pydoctor/commit/947c39d7c0453d3ed831269da2b196572127bed1) and [9e53d5b](https://github.com/twisted/pydoctor/commit/9e53d5b34cd26d0a41a10f269611b2d89fb2446f)

When it comes to graph generation, it seems there are different audiences:
- call and import graphs are useful for implementers
- class diagrams are useful for end users

I'm not sure what package diagrams as produced by epydoc looked like and for which audience they were intended.

Diagrams that are useful for end users can just be inserted into the output. Since all modern browsers support SVG, I'd prefer that over PNG. That would also make it easier to put hyperlinks in diagrams.

For diagrams that are intended for implementers, we could put them under the "show private API" filter. Another option would be to generate them to an entirely separate output document. Alternatively, we could export our code model and let external tools draw the graphs.

I do think we should be careful not to broaden our scope too much: the information we gather from source code is similar to what static code checkers do, so there is the temptation to add generic code quality features. But I think it's best for both usability and maintainability to keep focused on producing API documentation.

Call and import graphs are features where I'm not sure that they are all that strongly related to documentation.

> Also the documentation completeness checker [aa309cf](https://github.com/twisted/pydoctor/commit/aa309cf852946ab6b6eea2d647d5d9f6e9e2cdbe) It's a cool idea, and I'm sure there are people who will use it

I like the ability to have pydoctor report on the completeness of the documentation. But I think it would be better to have this as a an option on a normal run instead of a separate mode. We could add a "dry run" option to allow people to run pydoctor only as a doc linter, whether the completeness checking is enabled or not.

> The latex generator [9e53d5b](https://github.com/twisted/pydoctor/commit/9e53d5b34cd26d0a41a10f269611b2d89fb2446f) , this is to output the whole thing as PDF?

I can't find anything related to latex in that commit. Did you accidentally link the wrong commit or am I overlooking it?

Is it useful to have API docs as PDF? I think it would be less flexible than HTML as an electronic format and printed out documentation has no hyperlinks.

---
To summarize:
- I agree that some of these features would be useful to bring back
- bringing back the code may not always be the best approach to bringing back the feature

Since the scope is quite broad, I think it would be better to open individual issues for each features we want to bring back. That allows more focused discussion and we can close each issue when it's implemented.

**Comment by adiroiban** [comment]
-1 for call graph and PDF :)
And with PDF, it might be easier to create a print CSS and then convert from HTML+CSS to PDF.

-----

An "undocumented" report could be nice way to help developers identify what they have forgot to document.. but I think that it can be build on top of the current code.

**Comment by tristanlatr** [comment]
> Is it useful to have API docs as PDF? I think it would be less flexible than HTML as an electronic format and printed out documentation has no hyperlinks.

PDF can be searchable, the goal is to have eve thing in one document, that could be an extra HTLM page - it could contain everything. Then PDF converting would be easy.

> I'm not sure what package diagrams as produced by epydoc looked like and for which audience they were intended.

It seems that there are intended more for the users.

If you navigate the following docs, you will see class UML for each documented class, with the parent classes 

https://setupdocx.sourceforge.io/apiref/

And here is a link another exemple: http://epydoc.sourceforge.net/api/epydoc.apidoc.VariableDoc-class.html

EpyDoc supported the following features: 
     - ``.. digraph::``
     - epydoc <http://epydoc.sourceforge.net/api/epydoc.markup.restructuredtext-module.html#digraph_directive>
     - ``.. classtree::``
     - epydoc <http://epydoc.sourceforge.net/api/epydoc.markup.restructuredtext-module.html#classtree_directive>
     - ``.. packagetree::``
     - epydoc <http://epydoc.sourceforge.net/api/epydoc.markup.restructuredtext-module.html#package_directive>
     - ``.. importgraph::``
     - epydoc <http://epydoc.sourceforge.net/api/epydoc.markup.restructuredtext-module.html#importgraph_directive>
     - ``.. callgraph::``
     - epydoc <http://epydoc.sourceforge.net/api/epydoc.markup.restructuredtext-module.html#callgraph_directive> 

But also in a automatic way with 

```
--graph GRAPHTYPE
--
  | Include graphs of type GRAPHTYPE in the generated output. Graphs are generated using the Graphviz dot executable. 

If this executable is not on the path, then use --dotpath to specify its location. This option may be repeated to include multiple graph types in the output. To include all graphs, use --graph all. The available graph types are:

classtree: displays each class's base classes and subclasses;

callgraph: displays the callers and callees of each function or method. These graphs are based on profiling information, which must be specified using the --pstate option.

umlclass: displays each class's base classes and subclasses, using UML style. Methods and attributes are listed in the classes where they are defined. If type information is available about attributes (via the @type field), then those types are displayed as separate classes, and the attributes are displayed as associations.
```

While not everything seems to be useful or in the scope of pydoctor, maybe a few of these feature have a place nevertheless. I'm thinking of ``.. classtree::``, ``.. packagetree::`` and umlclass/classtree automatic diagrams. 

>  Alternatively, we could export our code model and let external tools draw the graphs.

This would be a cool idea, I guess it's discussed here https://github.com/twisted/pydoctor/issues/93 

Also that would give the possibility to design a markdown output writer as discussed here https://github.com/twisted/pydoctor/issues/80

**Comment by mthuurne** [comment]
> PDF can be searchable, the goal is to have eve thing in one document, that could be an extra HTLM page - it could contain everything. Then PDF converting would be easy.

A single-page HTML version of the documentation sounds a lot easier to implement than PDF output.

But if search is the goal, there might be better options. For example [hotdoc](https://github.com/hotdoc/hotdoc) has a client-side JavaScript search solution using a [trie](https://en.wikipedia.org/wiki/Trie). Using their code is probably not an option, both because of different licensing and because their index generator is written in C, but we could implement something similar.

> If you navigate the following docs, you will see class UML for each documented class, with the parent classes
> 
> https://setupdocx.sourceforge.io/apiref/

I'm not convinced that listing all methods in a box is clearer than listing them in a table. If we want to document inheritance, it might be better to show only the class name.

> And here is a link another exemple: http://epydoc.sourceforge.net/api/epydoc.apidoc.VariableDoc-class.html

This one looks a bit more useful, but the layout is not great here. Maybe modern versions of Graphviz perform better though.

In general, I've never been all too impressed with automatically generated UML diagrams. A generator doesn't know what information is relevant or not, so it will either include too much or too little. And I'm not sure that is solvable, since what is relevant depends on the question being asked and we don't know in advance what question the reader has when they are using the generated documentation.

That said, it would be useful to experiment a bit with different presentations, whether it is in text or diagrams. I'm sure the current presentation is far from optimal.

> > Alternatively, we could export our code model and let external tools draw the graphs.
> 
> This would be a cool idea, I guess it's discussed here #93

Yes, JSON sounds like a good format if we want to dump our data structures to file.

Maybe there is an existing standard for summarizing source code, if so we could consider that too.

> Also that would give the possibility to design a markdown output writer as discussed here #80

I'm not sure how useful that would be. That ticket mentions integration with mkdocs to get documentation on Read the Docs, but I think that improving our Sphinx integration might be an easier and yield better results. (RTD probably didn't support Sphinx yet in 2014, when the original ticket was filed.)

**Comment by tristanlatr** [comment]
OK, I'm going to close this ticket now. I'll open more specific tickets for specific features. 

More specific tickets or PR about re-enabling Epydoc features: #365 and #402
